plugins/org.jboss.tools.openshift.ui/src/org/jboss/tools/openshift/internal/ui/models/OpenShiftProjectCache.java

```diff
@@ -50,79 +52,119 @@ public OpenShiftProjectCache() {
 	}
 	
 	@Override
-	public synchronized Collection<IProjectAdapter> getProjectsFor(final IOpenShiftConnection conn) {
+	public Collection<IProjectAdapter> getProjectsFor(final IOpenShiftConnection conn) {
 		String key = getCacheKey(conn);
-		if(!cache.containsKey(key)) {
+		Set<IProjectAdapter> adapters = null;
```

Comment:
On the second glance it seems that this logic is not protected against concurrently creating 2 entries by two threads that call this method at approximately the same time and both passed !contains(key). Cache will finally have only one entry, but listeners will be notified of adapters created for both collections, in random order.
Possible code that fully protects this method against this case may be
boolean newAdapters = false;
Set&lt;IProjectAdapter> adapters = null;
synchronized(cache) {
    adapters = getEntry(key);
    if(adapters == null) {
        adapters = createEntry(key);
        newAdapters = true;
    }
}
if(newAdapters) {
    ....  //iterate projects.
}
However, that is not all. This method may be concurrently called with other methods that modify cache. Synchronization on cache and adapters prevents concurrent modification exceptions, but it does not save from listeners being notified of changes in an unpredictable order.